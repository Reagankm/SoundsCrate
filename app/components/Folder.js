/* @flow */
import React, { Component } from 'react';
import {
  Dimensions,
  Image,
  ScrollView,
  StyleSheet,
  Text,
  TouchableHighlight,
  View,
} from 'react-native';
import Record, { ANIMATION_DURATION } from './Record';
import constants from '../constants';
import _ from 'lodash';

const RECORD_HEIGHT_SPACE = 77;
const RECORD_TOP_SPACE = 70;
const FIRST_RECORD_TOP_SPACE = 10;
const SCROLL_THRESHOLD_MARGIN = 100;
const RECORD_SCROLL_DIFF_THRESHOLD = 162;

type Props = {
  records: Array<string>,
  onScrollEnd?: () => void
};

type State = {
  scrollDirection: string,
  recordShown: boolean,
  yPosition: number
};

export default class Folder extends Component {
  state: State = {
    scrollDirection: 'top',
    records: [],
    recordShown: false,
    yPosition: 0,
  };
  currentY: number = 0;
  scrollViewRef: ScrollView = null;
  debouncedOnScrollEnd: () => void;

  constructor(props: Props) {
    super(props);
    this.debouncedOnScrollEnd = _.debounce(this.props.onScrollEnd, 1000);
  }

  render() {
    const { records } = this.props;
    const height = records.length * RECORD_HEIGHT_SPACE;
    return (
      <ScrollView
        style={[styles.crate, { height }]}
        onScroll={event => this.onScroll(event)}
        ref={ref => this.scrollViewRef = ref}
        scrollEventThrottle={2}
        contentContainerStyle={[styles.container, { height }]}>
        {records.map((image, i) => {
          const zIndex = i + 1;
          const top = !i ? FIRST_RECORD_TOP_SPACE : i * RECORD_TOP_SPACE;
          return (<Record
            onShow={() => this.onRecordShow(top)}
            onHide={() => this.onRecordHide()}
            canBeShown={!this.state.recordShown}
            listYPosition={this.state.yPosition}
            maxYPosition={height}
            cover={image}
            isLast={i === records.length - 1}
            key={i}
            style={{ zIndex, top }}/>);
        })}
      </ScrollView>
    );
  }

  onScroll(event: any): void {
    const { contentOffset, contentSize, layoutMeasurement } = event.nativeEvent;
    const { y: yPosition } = contentOffset;
    const maxScrollHeight = contentSize.height - layoutMeasurement.height;
    const isBottomReached = yPosition + Dimensions.get('window').height >= contentSize.height;
    if (isBottomReached) {
      this.debouncedOnScrollEnd();
    }
    this.setState({ yPosition });
  }

  onRecordShow(topValue: number): void {
    const { yPosition } = this.state;
    const scrollTopDiff = (topValue - RECORD_SCROLL_DIFF_THRESHOLD) - yPosition;
    if (scrollTopDiff < 0) {
      setTimeout(() => {
        this.scrollViewRef.scrollTo({ y: yPosition + scrollTopDiff, animated: true });
      }, ANIMATION_DURATION);
    }
    this.setState({ recordShown: true });
  }

  onRecordHide(): void {
    this.setState({ recordShown: false });
  }
}

const styles = StyleSheet.create({
  container: {
    justifyContent: 'center',
    alignItems: 'flex-start',
    backgroundColor: constants.appBackgroundColor,
  },
})
