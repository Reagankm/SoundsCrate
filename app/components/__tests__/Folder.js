/* @flow */
import React from 'react';
import renderer from 'react-test-renderer';
import Folder from '../Folder';

jest.mock('../LoadingIndicator', () => 'LoadingIndicator');
jest.mock('../Record', () => ({
  __esModule: true,
  default: 'Record',
  ANIMATION_DURATION: 1,
}));
jest.mock('ScrollView', () => 'ScrollView');

describe('app/components/Folder', () => {
  it('should render', () => {
    const tree = renderer.create(
      <Folder
        records={[ 'foo', 'bar' ]}
        totalPages={1}
        isLoading={false}
        onScrollEnd={() => {}}
      />
    );
    expect(tree).toMatchSnapshot();
  });
  it('should render loading state if there are no records', () => {
    const tree = renderer.create(
      <Folder
        records={[]}
        totalPages={1}
        isLoading={false}
        onScrollEnd={() => {}}
      />
    );
    expect(tree).toMatchSnapshot();
  });
  it('should render loading state if prop passed', () => {
    const tree = renderer.create(
      <Folder
        records={[ 'foo', 'bar' ]}
        totalPages={1}
        isLoading={true}
        onScrollEnd={() => {}}
      />
    );
    expect(tree).toMatchSnapshot();
  });
  it('should render only one loading state if prop passed and there are no records', () => {
    const tree = renderer.create(
      <Folder
        records={[]}
        totalPages={1}
        isLoading={true}
        onScrollEnd={() => {}}
      />
    );
    expect(tree).toMatchSnapshot();
  });
});
