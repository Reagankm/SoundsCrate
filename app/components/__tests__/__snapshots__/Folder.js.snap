exports[`app/components/Folder should render 1`] = `
<View
  style={
    Object {
      "flex": 1,
    }
  }>
  <ScrollView
    contentContainerStyle={
      Array [
        Object {
          "alignItems": "flex-start",
          "backgroundColor": "#EEEEEE",
          "justifyContent": "center",
        },
        Object {
          "height": 154,
        },
      ]
    }
    onScroll={[Function]}
    scrollEventThrottle={2}
    style={
      Array [
        undefined,
        Object {
          "height": 154,
        },
      ]
    }>
    <Record
      canBeShown={true}
      cover="foo"
      isLast={false}
      listYPosition={0}
      maxYPosition={154}
      onHide={[Function]}
      onShow={[Function]}
      style={
        Object {
          "top": 10,
          "zIndex": 1,
        }
      } />
    <Record
      canBeShown={true}
      cover="bar"
      isLast={true}
      listYPosition={0}
      maxYPosition={154}
      onHide={[Function]}
      onShow={[Function]}
      style={
        Object {
          "top": 70,
          "zIndex": 2,
        }
      } />
  </ScrollView>
</View>
`;

exports[`app/components/Folder should render loading state if prop passed 1`] = `
<View
  style={
    Object {
      "flex": 1,
    }
  }>
  <ScrollView
    contentContainerStyle={
      Array [
        Object {
          "alignItems": "flex-start",
          "backgroundColor": "#EEEEEE",
          "justifyContent": "center",
        },
        Object {
          "height": 154,
        },
      ]
    }
    onScroll={[Function]}
    scrollEventThrottle={2}
    style={
      Array [
        undefined,
        Object {
          "height": 154,
        },
      ]
    }>
    <Record
      canBeShown={true}
      cover="foo"
      isLast={false}
      listYPosition={0}
      maxYPosition={154}
      onHide={[Function]}
      onShow={[Function]}
      style={
        Object {
          "top": 10,
          "zIndex": 1,
        }
      } />
    <Record
      canBeShown={true}
      cover="bar"
      isLast={true}
      listYPosition={0}
      maxYPosition={154}
      onHide={[Function]}
      onShow={[Function]}
      style={
        Object {
          "top": 70,
          "zIndex": 2,
        }
      } />
  </ScrollView>
  <LoadingIndicator
    isOverlay={true} />
</View>
`;

exports[`app/components/Folder should render loading state if there are no records 1`] = `
<View
  style={
    Object {
      "flex": 1,
    }
  }>
  <LoadingIndicator />
</View>
`;

exports[`app/components/Folder should render only one loading state if prop passed and there are no records 1`] = `
<View
  style={
    Object {
      "flex": 1,
    }
  }>
  <LoadingIndicator />
</View>
`;
